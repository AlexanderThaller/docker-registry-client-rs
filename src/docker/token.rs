use std::{
    collections::HashMap,
    sync::{
        Arc,
        RwLock,
    },
};

use chrono::{
    DateTime,
    Utc,
};
use reqwest::header::HeaderMap;
use serde::{
    Deserialize,
    Serialize,
};
use tracing::{
    info_span,
    Instrument,
};

#[cfg(feature = "redis_cache")]
use redis::AsyncCommands;

use crate::Image;

#[cfg(feature = "redis_cache")]
const REDIS_PREFIX: &str = "docker-registry-client:token";

#[derive(Debug, Eq, PartialEq, Hash)]
pub(super) struct CacheKey {
    image: Image,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub(super) struct Token {
    #[serde(rename = "token")]
    value: String,
    expires_in: Option<i64>,
    issued_at: Option<DateTime<Utc>>,
}

#[async_trait::async_trait]
pub(super) trait Cache: std::fmt::Debug + Send + Sync + dyn_clone::DynClone {
    async fn fetch(&self, key: &CacheKey) -> Option<Token>;
    async fn store(&self, key: CacheKey, token: Token);
}

dyn_clone::clone_trait_object!(Cache);

/// `NoCache` is a token cache that does not cache tokens.
#[derive(Debug, Default, Clone)]
pub(super) struct NoCache;

/// `MemoryTokenCache` is a token cache that caches tokens in memory.
#[derive(Debug, Default, Clone)]
pub(super) struct MemoryTokenCache {
    cache: Arc<RwLock<HashMap<CacheKey, Token>>>,
}

#[cfg(feature = "redis_cache")]
/// `RedisCache` is a token cache that caches tokens in Redis.
#[derive(Debug, Clone)]
pub(super) struct RedisCache {
    client: redis::Client,
}

impl std::fmt::Display for CacheKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let registry = self.image.registry.to_string();
        let namespace = self.image.namespace.as_ref();
        let repository = self.image.repository.as_ref();
        let image_name = &self.image.image_name.name;

        write!(f, "{registry}{namespace:?}{repository:?}{image_name}")
    }
}

#[async_trait::async_trait]
impl Cache for NoCache {
    async fn fetch(&self, _key: &CacheKey) -> Option<Token> {
        None
    }

    async fn store(&self, _key: CacheKey, _token: Token) {}
}

#[async_trait::async_trait]
impl Cache for MemoryTokenCache {
    #[tracing::instrument]
    async fn fetch(&self, key: &CacheKey) -> Option<Token> {
        self.cache
            .read()
            .expect("failed to get read lock")
            .get(key)
            .cloned()
            .and_then(|token| {
                if let Some(expires_in) = token.expires_in {
                    token
                        .issued_at
                        .map(|issued_at| issued_at + chrono::Duration::seconds(expires_in))
                        .and_then(|expires_at| {
                            if expires_at < Utc::now() {
                                None
                            } else {
                                Some(token)
                            }
                        })
                } else {
                    Some(token)
                }
            })
    }

    #[tracing::instrument]
    async fn store(&self, key: CacheKey, token: Token) {
        self.cache
            .write()
            .expect("failed to get write lock")
            .insert(key, token);
    }
}

impl From<&Image> for CacheKey {
    fn from(image: &Image) -> Self {
        Self {
            image: image.clone(),
        }
    }
}

impl TryInto<HeaderMap> for Token {
    type Error = reqwest::header::InvalidHeaderValue;

    fn try_into(self) -> Result<HeaderMap, Self::Error> {
        let mut headers = HeaderMap::new();
        headers.insert("Authorization", format!("Bearer {}", self.value).parse()?);

        Ok(headers)
    }
}

#[cfg(feature = "redis_cache")]
impl RedisCache {
    #[must_use]
    pub fn new(client: redis::Client) -> Self {
        Self { client }
    }
}

#[cfg(feature = "redis_cache")]
#[async_trait::async_trait]
impl Cache for RedisCache {
    #[tracing::instrument]
    async fn fetch(&self, key: &CacheKey) -> Option<Token> {
        let mut connection = self
            .client
            .get_multiplexed_async_connection()
            .instrument(info_span!("get redis connection"))
            .await
            .expect("failed to get connection");

        let key = format!("{REDIS_PREFIX}:{key}");

        let exists: bool = connection
            .exists(&key)
            .instrument(info_span!("check if key exists"))
            .await
            .expect("failed to check if key exists");

        if !exists {
            return None;
        }

        let value: String = connection
            .get(&key)
            .instrument(info_span!("get value"))
            .await
            .expect("failed to get value");

        let token = serde_json::from_str(&value).expect("failed to deserialize token");

        Some(token)
    }

    #[tracing::instrument]
    async fn store(&self, key: CacheKey, token: Token) {
        let mut connection = self
            .client
            .get_multiplexed_async_connection()
            .instrument(info_span!("get redis connection"))
            .await
            .expect("failed to get connection");

        let key = format!("{REDIS_PREFIX}:{key}");

        let value = serde_json::to_string(&token).expect("failed to serialize token");

        connection
            .set::<&String, String, String>(&key, value)
            .instrument(info_span!("set value"))
            .await
            .expect("failed to set value");

        if let Some(expires_in) = token.expires_in {
            connection
                .expire::<&String, String>(&key, expires_in)
                .instrument(info_span!("set expire"))
                .await
                .expect("failed to set expiration");
        }
    }
}

#[cfg(test)]
mod tests {
    mod token {
        mod deserialize {
            use crate::docker::Token;

            #[test]
            fn github() {
                const INPUT: &str =
                    r#"{"token":"djE6c2lnc3RvcmUvY29zaWduL2Nvc2lnbjoxNzI1NDM2OTAwNTczODMyMzM2"}"#;

                let got = serde_json::from_str::<Token>(INPUT).unwrap();
                insta::assert_json_snapshot!(got);
            }

            #[test]
            fn dockerhub() {
                const INPUT: &str = r#"{"token":"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsIng1YyI6WyJNSUlFRmpDQ0F2NmdBd0lCQWdJVVZOajJRbU1JWnUzeGl0NUJ1RTlvRWdoVU5KUXdEUVlKS29aSWh2Y05BUUVMQlFBd2dZWXhDekFKQmdOVkJBWVRBbFZUTVJNd0VRWURWUVFJRXdwRFlXeHBabTl5Ym1saE1SSXdFQVlEVlFRSEV3bFFZV3h2SUVGc2RHOHhGVEFUQmdOVkJBb1RERVJ2WTJ0bGNpd2dTVzVqTGpFVU1CSUdBMVVFQ3hNTFJXNW5hVzVsWlhKcGJtY3hJVEFmQmdOVkJBTVRHRVJ2WTJ0bGNpd2dTVzVqTGlCRmJtY2dVbTl2ZENCRFFUQWVGdzB5TkRBeE1UWXdOak0yTURCYUZ3MHlOVEF4TVRVd05qTTJNREJhTUlHRk1Rc3dDUVlEVlFRR0V3SlZVekVUTUJFR0ExVUVDQk1LUTJGc2FXWnZjbTVwWVRFU01CQUdBMVVFQnhNSlVHRnNieUJCYkhSdk1SVXdFd1lEVlFRS0V3eEViMk5yWlhJc0lFbHVZeTR4RkRBU0JnTlZCQXNUQzBWdVoybHVaV1Z5YVc1bk1TQXdIZ1lEVlFRREV4ZEViMk5yWlhJc0lFbHVZeTRnUlc1bklFcFhWQ0JEUVRDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBTWI4eHR6ZDQ1UWdYekV0bWMxUEJsdWNGUnlzSUF4UUJCN3lSNjdJemdMd05IS24rbUdKTzV5alh6amtLZm5zWm1JRURnZFlraEpBbGNYYTdQa1BFaCtqcTRGNWNaaWtkTmFUQmM3alNkTFJzTVlVa3dwWTl4WUVqYitCYnVGUWVxa0R2RXNqbFJJTzRQK0FsRlhNMDhMYlpIZ3hFWUdkbFk3WFlhT1BLMmE1aUd2eVFRb09GVmZjZDd2ekhaREVBMHZqVmU1M0xLdjVMYmh6TzcxZHRxS0RwNEhnVWR5N1pENDFNN3I1bTd5eE1LeFNpQmJHZTFvem5Wamh1ck5GNHdGSml5bVU4YkhTV2tVTXVLQ3JTbEd4d1NCZFVZNDRyaEh2UW5zYmgzUFF2TUZTWTQ4REdoNFhUUldjSzFWUVlSTnA2ZWFXUVg1RUpJSXVJbjJQOVBzQ0F3RUFBYU43TUhrd0RnWURWUjBQQVFIL0JBUURBZ0dtTUJNR0ExVWRKUVFNTUFvR0NDc0dBUVVGQndNQk1CSUdBMVVkRXdFQi93UUlNQVlCQWY4Q0FRQXdIUVlEVlIwT0JCWUVGSnVRYXZTZHVScm5kRXhLTTAwV2Z2czh5T0RaTUI4R0ExVWRJd1FZTUJhQUZGSGVwRE9ZQ0Y5Qnc5dXNsY0RVUW5CalU3MS9NQTBHQ1NxR1NJYjNEUUVCQ3dVQUE0SUJBUUNDWW0xUVorUUZ1RVhkSWpiNkg4bXNyVFBRSlNnR0JpWDFXSC9QRnpqZlJGeHc3dTdDazBRb0FXZVNqV3JWQWtlVlZQN3J2REpwZ0ZoeUljdzNzMXRPVjN0OGp3cXJTUmc2R285dUd2OG9IWUlLTm9YaDErUFVDTG44b0JwYUJsOUxzSWtsL2FHMG9lY0hrcDVLYmtBNjN6eTFxSUVXNFAzWVJLSk9hSGoxYWFiOXJLc3lRSHF6SUl4TnlDRVVINTMwU1B4RUNMbE53YWVKTDVmNXIxUW5wSi9GM3Q5Vk8xZ0Y2RFpiNitPczdTV29ocGhWZlRCOERkL1VjSk1VOGp2YlF3MWRVREkwelNEdXo2aHNJbGdITk0yak04M0lOS1VqNjNaRDMwRG15ejQvczFFdGgyQmlKK2RHdnFpQkRzaWhaR0tyQnJzUzhWVkRBd3hDeDVRMyJdfQ.eyJhY2Nlc3MiOlt7ImFjdGlvbnMiOlsicHVsbCJdLCJuYW1lIjoibGlicmFyeS9hbHBpbmUiLCJwYXJhbWV0ZXJzIjp7InB1bGxfbGltaXQiOiIxMDAiLCJwdWxsX2xpbWl0X2ludGVydmFsIjoiMjE2MDAifSwidHlwZSI6InJlcG9zaXRvcnkifV0sImF1ZCI6InJlZ2lzdHJ5LmRvY2tlci5pbyIsImV4cCI6MTcyNTQzNzIwMSwiaWF0IjoxNzI1NDM2OTAxLCJpc3MiOiJhdXRoLmRvY2tlci5pbyIsImp0aSI6ImRja3JfanRpX1dVVnQ2RldmdGRyNUU4UUNTTmwxb1M3NG1LUT0iLCJuYmYiOjE3MjU0MzY2MDEsInN1YiI6IiJ9.SMcP5s3bLnOmC0X3pG4b9N8PD3Dtpk5xovCzdcje-79O-lRx0Tqa2zruCauMENetmmq2CoxZOsY1ostvrAfZaDp2f_KkjaftymHhvgx60ppxIZuHIwZJjt9v7XprJuK8OZ1QxHL-HKp5nw8dRwa-IyScs-YevTz7PMD5B7uy36AZs1db2FwT_6Ygzhkvuq3OKgr70Mie2WHAoyShEJs7V6b9tywiSXPEgu2Nf7VSrj6PeZOi__IQl0wXo7w4j7C8o6ijUAotPRfMoQUXN1j9_4ql-yJZpp5jVmLsEOaOVKPjoIvnyDWYs9KVluUF4NKgvOydgIP-MeccXXIFOg4bhA","access_token":"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsIng1YyI6WyJNSUlFRmpDQ0F2NmdBd0lCQWdJVVZOajJRbU1JWnUzeGl0NUJ1RTlvRWdoVU5KUXdEUVlKS29aSWh2Y05BUUVMQlFBd2dZWXhDekFKQmdOVkJBWVRBbFZUTVJNd0VRWURWUVFJRXdwRFlXeHBabTl5Ym1saE1SSXdFQVlEVlFRSEV3bFFZV3h2SUVGc2RHOHhGVEFUQmdOVkJBb1RERVJ2WTJ0bGNpd2dTVzVqTGpFVU1CSUdBMVVFQ3hNTFJXNW5hVzVsWlhKcGJtY3hJVEFmQmdOVkJBTVRHRVJ2WTJ0bGNpd2dTVzVqTGlCRmJtY2dVbTl2ZENCRFFUQWVGdzB5TkRBeE1UWXdOak0yTURCYUZ3MHlOVEF4TVRVd05qTTJNREJhTUlHRk1Rc3dDUVlEVlFRR0V3SlZVekVUTUJFR0ExVUVDQk1LUTJGc2FXWnZjbTVwWVRFU01CQUdBMVVFQnhNSlVHRnNieUJCYkhSdk1SVXdFd1lEVlFRS0V3eEViMk5yWlhJc0lFbHVZeTR4RkRBU0JnTlZCQXNUQzBWdVoybHVaV1Z5YVc1bk1TQXdIZ1lEVlFRREV4ZEViMk5yWlhJc0lFbHVZeTRnUlc1bklFcFhWQ0JEUVRDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBTWI4eHR6ZDQ1UWdYekV0bWMxUEJsdWNGUnlzSUF4UUJCN3lSNjdJemdMd05IS24rbUdKTzV5alh6amtLZm5zWm1JRURnZFlraEpBbGNYYTdQa1BFaCtqcTRGNWNaaWtkTmFUQmM3alNkTFJzTVlVa3dwWTl4WUVqYitCYnVGUWVxa0R2RXNqbFJJTzRQK0FsRlhNMDhMYlpIZ3hFWUdkbFk3WFlhT1BLMmE1aUd2eVFRb09GVmZjZDd2ekhaREVBMHZqVmU1M0xLdjVMYmh6TzcxZHRxS0RwNEhnVWR5N1pENDFNN3I1bTd5eE1LeFNpQmJHZTFvem5Wamh1ck5GNHdGSml5bVU4YkhTV2tVTXVLQ3JTbEd4d1NCZFVZNDRyaEh2UW5zYmgzUFF2TUZTWTQ4REdoNFhUUldjSzFWUVlSTnA2ZWFXUVg1RUpJSXVJbjJQOVBzQ0F3RUFBYU43TUhrd0RnWURWUjBQQVFIL0JBUURBZ0dtTUJNR0ExVWRKUVFNTUFvR0NDc0dBUVVGQndNQk1CSUdBMVVkRXdFQi93UUlNQVlCQWY4Q0FRQXdIUVlEVlIwT0JCWUVGSnVRYXZTZHVScm5kRXhLTTAwV2Z2czh5T0RaTUI4R0ExVWRJd1FZTUJhQUZGSGVwRE9ZQ0Y5Qnc5dXNsY0RVUW5CalU3MS9NQTBHQ1NxR1NJYjNEUUVCQ3dVQUE0SUJBUUNDWW0xUVorUUZ1RVhkSWpiNkg4bXNyVFBRSlNnR0JpWDFXSC9QRnpqZlJGeHc3dTdDazBRb0FXZVNqV3JWQWtlVlZQN3J2REpwZ0ZoeUljdzNzMXRPVjN0OGp3cXJTUmc2R285dUd2OG9IWUlLTm9YaDErUFVDTG44b0JwYUJsOUxzSWtsL2FHMG9lY0hrcDVLYmtBNjN6eTFxSUVXNFAzWVJLSk9hSGoxYWFiOXJLc3lRSHF6SUl4TnlDRVVINTMwU1B4RUNMbE53YWVKTDVmNXIxUW5wSi9GM3Q5Vk8xZ0Y2RFpiNitPczdTV29ocGhWZlRCOERkL1VjSk1VOGp2YlF3MWRVREkwelNEdXo2aHNJbGdITk0yak04M0lOS1VqNjNaRDMwRG15ejQvczFFdGgyQmlKK2RHdnFpQkRzaWhaR0tyQnJzUzhWVkRBd3hDeDVRMyJdfQ.eyJhY2Nlc3MiOlt7ImFjdGlvbnMiOlsicHVsbCJdLCJuYW1lIjoibGlicmFyeS9hbHBpbmUiLCJwYXJhbWV0ZXJzIjp7InB1bGxfbGltaXQiOiIxMDAiLCJwdWxsX2xpbWl0X2ludGVydmFsIjoiMjE2MDAifSwidHlwZSI6InJlcG9zaXRvcnkifV0sImF1ZCI6InJlZ2lzdHJ5LmRvY2tlci5pbyIsImV4cCI6MTcyNTQzNzIwMSwiaWF0IjoxNzI1NDM2OTAxLCJpc3MiOiJhdXRoLmRvY2tlci5pbyIsImp0aSI6ImRja3JfanRpX1dVVnQ2RldmdGRyNUU4UUNTTmwxb1M3NG1LUT0iLCJuYmYiOjE3MjU0MzY2MDEsInN1YiI6IiJ9.SMcP5s3bLnOmC0X3pG4b9N8PD3Dtpk5xovCzdcje-79O-lRx0Tqa2zruCauMENetmmq2CoxZOsY1ostvrAfZaDp2f_KkjaftymHhvgx60ppxIZuHIwZJjt9v7XprJuK8OZ1QxHL-HKp5nw8dRwa-IyScs-YevTz7PMD5B7uy36AZs1db2FwT_6Ygzhkvuq3OKgr70Mie2WHAoyShEJs7V6b9tywiSXPEgu2Nf7VSrj6PeZOi__IQl0wXo7w4j7C8o6ijUAotPRfMoQUXN1j9_4ql-yJZpp5jVmLsEOaOVKPjoIvnyDWYs9KVluUF4NKgvOydgIP-MeccXXIFOg4bhA","expires_in":300,"issued_at":"2024-09-04T08:01:41.048681016Z"}"#;

                let got = serde_json::from_str::<Token>(INPUT).unwrap();
                insta::assert_json_snapshot!(got);
            }
        }
    }
}
