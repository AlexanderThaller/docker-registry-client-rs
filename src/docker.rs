use reqwest::{
    header::HeaderMap,
    Client as HTTPClient,
};
use serde::{
    Deserialize,
    Serialize,
};
use tracing::{
    info_span,
    Instrument,
};
use url::Url;

use crate::{
    Image,
    Manifest,
    Registry,
};

mod error;
pub mod token;
pub mod token_cache;

pub use error::Error;
use token::Token;
use token_cache::Cache as TokenCache;

#[derive(Debug, Clone)]
pub struct Client {
    client: HTTPClient,
    token_cache: Box<dyn TokenCache + Send>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Response {
    pub digest: Option<String>,
    pub manifest: Manifest,
}

impl Default for Client {
    fn default() -> Self {
        Self {
            client: HTTPClient::new(),
            token_cache: Box::new(token_cache::MemoryTokenCache::default()),
        }
    }
}

impl Client {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    pub fn set_cache_memory(&mut self) {
        self.token_cache = Box::new(token_cache::MemoryTokenCache::default());
    }

    pub fn disable_caching(&mut self) {
        self.token_cache = Box::new(token_cache::NoCache);
    }

    #[cfg(feature = "redis_cache")]
    pub fn set_cache_redis(&mut self, redis_client: redis::Client) {
        self.token_cache = Box::new(token_cache::RedisCache::new(redis_client));
    }

    #[tracing::instrument]
    pub async fn get_manifest_url(&self, url: &Url, image: &Image) -> Result<Response, Error> {
        let mut headers = self.get_headers(image).await?;

        let accept_header = [
            "application/vnd.docker.container.image.v1+json",
            "application/vnd.docker.distribution.manifest.list.v2+json",
            "application/vnd.docker.distribution.manifest.v2+json",
            "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "application/vnd.docker.image.rootfs.foreign.diff.tar.gzip",
            "application/vnd.docker.plugin.v1+json",
            "application/vnd.oci.image.index.v1+json",
            "application/vnd.oci.image.manifest.v1+json",
        ]
        .join(", ");

        headers.insert(
            "Accept",
            accept_header
                .parse()
                .map_err(Error::ParseManifestAcceptHeader)?,
        );

        let response = self
            .client
            .get(url.as_str())
            .headers(headers)
            .send()
            .instrument(info_span!("get manifest request"))
            .await
            .map_err(Error::GetManifest)?;

        let status = response.status();

        let digest = response
            .headers()
            .get("Docker-Content-Digest")
            .map(|header| {
                header
                    .to_str()
                    .map(String::from)
                    .map_err(Error::ParseDockerContentDigestHeader)
            })
            .transpose()?;

        let body = response
            .text()
            .instrument(info_span!("extract manifest request body"))
            .await
            .map_err(Error::ExtractManifestBody)?;

        if !status.is_success() {
            if status == reqwest::StatusCode::NOT_FOUND {
                return Err(Error::ManifestNotFound(url.clone()));
            }

            return Err(Error::FailedManifestRequest(status, body));
        }

        let manifest =
            serde_json::from_str(&body).map_err(|e| Error::DeserializeManifestBody(e, body))?;

        Ok(Response { digest, manifest })
    }

    /// # Errors
    /// Returns an error if the request fails.
    /// Returns an error if the response body is not valid JSON.
    /// Returns an error if the response body is not a valid manifest.
    /// Returns an error if the response status is not successful.
    #[tracing::instrument(skip_all)]
    pub async fn get_manifest(&self, image: &Image) -> Result<Response, Error> {
        let registry_domain = image.registry.registry_domain();

        let url = Url::parse(&format!(
            "https://{domain}/v2/{namespace}{repository}{image_name}/manifests/{identifier}",
            domain = registry_domain,
            namespace = match image.namespace {
                Some(ref namespace) => format!("{namespace}/"),
                None => String::new(),
            },
            repository = match image.repository {
                Some(ref repository) => format!("{repository}/"),
                None => String::new(),
            },
            image_name = image.image_name.name,
            identifier = image.image_name.identifier
        ))
        .map_err(Error::InvalidManifestUrl)?;

        self.get_manifest_url(&url, image).await
    }

    #[tracing::instrument(skip_all)]
    async fn get_headers(&self, image: &Image) -> Result<HeaderMap, Error> {
        if !image.registry.needs_authentication() {
            return Ok(HeaderMap::new());
        }

        let cache_key = image.into();

        let token = self
            .token_cache
            .fetch(&cache_key)
            .await
            .map_err(Error::FetchToken)?;

        let token = if let Some(token) = token {
            token
        } else {
            let namespace = match &image.namespace {
                Some(namespace) => format!("{namespace}/"),
                None => String::new(),
            };

            let repository = match &image.repository {
                Some(repository) => format!("{repository}/"),
                None => String::new(),
            };

            let token_url = match image.registry {
                Registry::Github => format!(
                    "https://ghcr.io/token?scope=repository:{namespace}{repository}{image_name}:pull&service=ghcr.io",
                    image_name = image.image_name.name
                ),

                Registry::DockerHub => format!("https://auth.docker.io/token?service=registry.docker.io&scope=repository:{namespace}{repository}{image_name}:pull&service=registry.docker.io", image_name = image.image_name.name),

                Registry::Quay => format!("https://quay.io/v2/auth?scope=repository:{namespace}{repository}{image_name}:pull&service=quay.io", image_name = image.image_name.name),

                Registry::RedHat | Registry::K8s | Registry::Google | Registry::Microsoft => return Ok(HeaderMap::new()),
            };

            let token_url = Url::parse(&token_url).map_err(Error::InvalidTokenUrl)?;

            let response = self
                .client
                .get(token_url)
                .send()
                .instrument(info_span!("get token request"))
                .await
                .map_err(Error::GetToken)?;

            let body = response
                .text()
                .instrument(info_span!("extract token request body"))
                .await
                .map_err(Error::ExtractTokenBody)?;

            let token: Token =
                serde_json::from_str(&body).map_err(|e| Error::DeserializeToken(e, body))?;

            self.token_cache
                .store(cache_key, token.clone())
                .await
                .map_err(Error::StoreToken)?;

            token
        };

        let headers = token.try_into().map_err(Error::ParseAuthorizationHeader)?;

        Ok(headers)
    }
}

#[cfg(test)]
#[expect(clippy::unwrap_used, reason = "using unwrap in tests is fine")]
mod tests {
    mod dockerhub {
        use crate::{
            Client,
            Image,
            ImageName,
            Registry,
            Tag,
        };
        use either::Either;

        #[tokio::test]
        async fn alpine() {
            let client = Client::new();

            let image_name = Image {
                registry: Registry::DockerHub,
                namespace: None,
                repository: Some("library".to_string()),
                image_name: ImageName {
                    name: "alpine".to_string(),
                    identifier: Either::Left(Tag::Specific("3.20".to_string())),
                },
            };

            let response = client.get_manifest(&image_name).await.unwrap();

            insta::assert_json_snapshot!(response);
        }
    }

    mod redhat {
        use crate::{
            Client,
            Image,
            ImageName,
            Registry,
            Tag,
        };
        use either::Either;

        #[tokio::test]
        async fn ubi8() {
            let client = Client::new();

            let image = Image {
                registry: Registry::RedHat,
                namespace: None,
                repository: None,
                image_name: ImageName {
                    name: "ubi8".to_string(),
                    identifier: Either::Left(Tag::Specific("8.9".to_string())),
                },
            };

            let response = client.get_manifest(&image).await.unwrap();

            insta::assert_json_snapshot!(response);
        }

        #[tokio::test]
        async fn cosign() {
            const INPUT: &str = "ghcr.io/sigstore/cosign/cosign:v2.4.0";

            let client = Client::new();
            let image = INPUT.parse().unwrap();
            let response = client.get_manifest(&image).await.unwrap();

            insta::assert_json_snapshot!(response);
        }

        #[tokio::test]
        async fn playwright() {
            const INPUT: &str = "mcr.microsoft.com/playwright:v1.48.2-noble";

            let client = Client::new();
            let image = INPUT.parse().unwrap();
            let response = client.get_manifest(&image).await.unwrap();

            insta::assert_json_snapshot!(response);
        }
    }
}
