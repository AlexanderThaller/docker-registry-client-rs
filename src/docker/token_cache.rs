use std::{
    collections::HashMap,
    sync::Arc,
};

use chrono::Utc;
use tokio::sync::RwLock;
use tracing::{
    Instrument,
    info_span,
};

use crate::docker::token::{
    CacheKey,
    Token,
};

#[cfg(feature = "redis_cache")]
use redis::AsyncCommands;

#[cfg(feature = "redis_cache")]
const REDIS_PREFIX: &str = "docker-registry-client:token";

#[derive(Debug)]
pub enum FetchError {
    CheckExists(redis::RedisError),
    DeserializeToken(serde_json::Error),
    GetConnection(redis::RedisError),
    GetValue(redis::RedisError),
}

#[derive(Debug)]
pub enum StoreError {
    GetConnection(redis::RedisError),
    SerializeToken(serde_json::Error),
    SetExpiration(redis::RedisError),
    SetValue(redis::RedisError),
}

#[async_trait::async_trait]
pub(super) trait Cache: std::fmt::Debug + Send + Sync + dyn_clone::DynClone {
    async fn fetch(&self, key: &CacheKey) -> Result<Option<Token>, FetchError>;
    async fn store(&self, key: CacheKey, token: Token) -> Result<(), StoreError>;
}

dyn_clone::clone_trait_object!(Cache);

/// `NoCache` is a token cache that does not cache tokens.
#[derive(Debug, Default, Clone)]
pub(super) struct NoCache;

/// `MemoryTokenCache` is a token cache that caches tokens in memory.
#[derive(Debug, Default, Clone)]
pub(super) struct MemoryTokenCache {
    cache: Arc<RwLock<HashMap<CacheKey, Token>>>,
}

#[cfg(feature = "redis_cache")]
/// `RedisCache` is a token cache that caches tokens in Redis.
#[derive(Debug, Clone)]
pub(super) struct RedisCache {
    client: redis::Client,
}

impl std::fmt::Display for FetchError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::CheckExists(e) => write!(f, "failed to check if key exists: {e}"),
            Self::DeserializeToken(e) => write!(f, "failed to deserialize token: {e}"),
            Self::GetConnection(e) => write!(f, "failed to get redis connection: {e}"),
            Self::GetValue(e) => write!(f, "failed to get value from redis: {e}"),
        }
    }
}

impl std::error::Error for FetchError {}

impl std::fmt::Display for StoreError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::GetConnection(e) => write!(f, "failed to get redis connection: {e}"),
            Self::SerializeToken(e) => write!(f, "failed to serialize token: {e}"),
            Self::SetExpiration(e) => write!(f, "failed to set expiration: {e}"),
            Self::SetValue(e) => write!(f, "failed to set value in redis: {e}"),
        }
    }
}

impl std::error::Error for StoreError {}

#[async_trait::async_trait]
impl Cache for NoCache {
    async fn fetch(&self, _key: &CacheKey) -> Result<Option<Token>, FetchError> {
        Ok(None)
    }

    async fn store(&self, _key: CacheKey, _token: Token) -> Result<(), StoreError> {
        Ok(())
    }
}

#[async_trait::async_trait]
impl Cache for MemoryTokenCache {
    #[tracing::instrument]
    async fn fetch(&self, key: &CacheKey) -> Result<Option<Token>, FetchError> {
        let result = self.cache.read().await.get(key).cloned().and_then(|token| {
            if let Some(expires_in) = token.expires_in {
                token
                    .issued_at
                    .map(|issued_at| issued_at + chrono::Duration::seconds(expires_in))
                    .and_then(|expires_at| {
                        if expires_at < Utc::now() {
                            None
                        } else {
                            Some(token)
                        }
                    })
            } else {
                Some(token)
            }
        });

        Ok(result)
    }

    #[tracing::instrument]
    async fn store(&self, key: CacheKey, token: Token) -> Result<(), StoreError> {
        self.cache.write().await.insert(key, token);

        Ok(())
    }
}

#[cfg(feature = "redis_cache")]
impl RedisCache {
    #[must_use]
    pub fn new(client: redis::Client) -> Self {
        Self { client }
    }
}

#[cfg(feature = "redis_cache")]
#[async_trait::async_trait]
impl Cache for RedisCache {
    #[tracing::instrument]
    async fn fetch(&self, key: &CacheKey) -> Result<Option<Token>, FetchError> {
        let mut connection = self
            .client
            .get_multiplexed_async_connection()
            .instrument(info_span!("get redis connection"))
            .await
            .map_err(FetchError::GetConnection)?;

        let key = format!("{REDIS_PREFIX}:{key}");

        let exists: bool = connection
            .exists(&key)
            .instrument(info_span!("check if key exists"))
            .await
            .map_err(FetchError::CheckExists)?;

        if !exists {
            return Ok(None);
        }

        let value: String = connection
            .get(&key)
            .instrument(info_span!("get value"))
            .await
            .map_err(FetchError::GetValue)?;

        let token = serde_json::from_str(&value).map_err(FetchError::DeserializeToken)?;

        Ok(Some(token))
    }

    #[tracing::instrument]
    async fn store(&self, key: CacheKey, token: Token) -> Result<(), StoreError> {
        let mut connection = self
            .client
            .get_multiplexed_async_connection()
            .instrument(info_span!("get redis connection"))
            .await
            .map_err(StoreError::GetConnection)?;

        let key = format!("{REDIS_PREFIX}:{key}");

        let value = serde_json::to_string(&token).map_err(StoreError::SerializeToken)?;

        connection
            .set::<&String, String, String>(&key, value)
            .instrument(info_span!("set value"))
            .await
            .map_err(StoreError::SetValue)?;

        if let Some(expires_in) = token.expires_in {
            connection
                .expire::<&String, String>(&key, expires_in)
                .instrument(info_span!("set expire"))
                .await
                .map_err(StoreError::SetExpiration)?;
        }

        Ok(())
    }
}
