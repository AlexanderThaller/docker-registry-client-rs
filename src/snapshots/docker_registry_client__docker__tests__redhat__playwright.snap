---
source: src/docker.rs
expression: response
---
{
  "digest": "sha256:c43809dabac73ac46b136409daa0a7d5411fb3ccc1e7d0fad1be5383a7a0f6ef",
  "manifest": {
    "schemaVersion": 2,
    "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
    "manifests": [
      {
        "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
        "size": 1165,
        "digest": "sha256:0ca8187c00ef90f9fc10e3a44585f7e996606e4d4120ad79198a7d5c338b62bb",
        "platform": {
          "architecture": "amd64",
          "os": "linux"
        }
      },
      {
        "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
        "size": 1165,
        "digest": "sha256:ab354aac05ed5201bc6a2ca332942cbd981d29da950d001c9e5c6191849c2738",
        "platform": {
          "architecture": "arm64",
          "os": "linux"
        }
      }
    ]
  }
}
